lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -383,6 +433,811 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
+  return It->second;
+}
+
+static UpcastablePointer<model::FunctionEdge>
+makeEdge(MetaAddress Destination, model::FunctionEdgeType::Values Type) {
+  model::FunctionEdge *Result = nullptr;
+  if (model::FunctionEdgeType::isCall(Type))
+    Result = new model::CallEdge(Destination, Type);
+  else
+    Result = new model::FunctionEdge(Destination, Type);
+  return UpcastablePointer<model::FunctionEdge>(Result);
+};
+
+template<class FunctionOracle>
+SortedVector<model::BasicBlock>
+CFEPAnalyzer<FunctionOracle>::collectDirectCFG(llvm::BasicBlock *Entry,
+                                               llvm::Function *F) {
+  using namespace llvm;
+  SortedVector<model::BasicBlock> CFG;
+
+  for (auto &BB : *F) {
+    if (GCBI::isJumpTarget(&BB)) {
+      auto Start = getBasicBlockPC(&BB);
+      MetaAddress End;
+      uint64_t Size;
+
+      bool Stop = false;
+      model::BasicBlock Block{ Start };
+      OnceQueue<BasicBlock *> Queue;
+      Queue.insert(&BB);
+
+      while (!Queue.empty()) {
+        BasicBlock *Current = Queue.pop();
+
+        auto [CandidateEnd, CandidateSize] = getPC(Current->getTerminator());
+        if (CandidateEnd > End) {
+          End = CandidateEnd;
+          Size = CandidateSize;
+        }
+
+        for (auto *Succ : successors(Current)) {
+          if (GCBI::isJumpTarget(Succ)) {
+            auto Destination = getBasicBlockPC(Succ);
+            auto Edge = makeEdge(Destination,
+                                 model::FunctionEdgeType::DirectBranch);
+            Block.Successors.insert(Edge);
+          } else if (UnexpectedPCCloned == Succ && succ_size(Current) == 1) {
+            Block.Successors.insert(makeEdge(MetaAddress::invalid(),
+                                             model::FunctionEdgeType::LongJmp));
+          } else {
+            auto *I = Succ->getFirstNonPHI();
+
+            if (isa<CallInst>(I)
+                && cast<CallInst>(I)->getCalledFunction() == PreHookMarker) {
+              MetaAddress Destination;
+              model::FunctionEdgeType::Values Type;
+              auto *CalleePC = cast<CallInst>(I)->getArgOperand(1);
+
+              // Direct or indirect call?
+              if (isa<ConstantStruct>(CalleePC)) {
+                Destination = MetaAddress::fromConstant(CalleePC);
+                Type = model::FunctionEdgeType::FunctionCall;
+              } else {
+                Destination = MetaAddress::invalid();
+                Type = model::FunctionEdgeType::IndirectCall;
+              }
+
+              auto Edge = makeEdge(Destination, Type);
+              auto DestTy = Oracle.getFunctionType(Destination);
+              if (DestTy == FunctionTypeValue::NoReturn)
+                cast<model::CallEdge>(Edge.get())->IsNoReturn = true;
+
+              Block.Successors.insert(Edge);
+            } else if (auto *Call = getCallTo(I, "function_call")) {
+              // Marker can be erased now, no longer needed.
+              Call->eraseFromParent();
```

Comment:
This is a bit suspicious.
This function should only collect information, not make changes.
In fact, we could even have the function passed in as `const`.
Can you move this in a more appropriate location?

Reply:
We decided to use `function_call` to also mark the presence of fake functions, without introducing another marker (e.g., `call void @fake_function_marker`). The marker was already erased after integrating the function callee's summary, thus it is completely safe to erase it at this stage, it won't be of no further use.

Reply:
As you prefer, but `analyze` is so pretty right now, just function calls.